### ISSUE TITLE
[question] x-build cross compiler

### ISSUE THREAD

**Issue by mjvankampen** [conan-io/conan#6971]
I am still a bit unsure how to use the x-build setup (with host and build profile) and cross-compilers.

Currently I have the following: a cross-compiler package with multiple settings (and thus detected as separate packages):
os: Neutrino
arch: x86/x86_64/armv7/armv8
compiler: qcc
os_build: Linux/Macos/Windows
os_arch: x86_64

In total this means 12 packages.

If I use the new setup I can only define 4 packages (for arch x86/x86_64/armv7/armv8), as I cannot separate between windows, linux and macos. Is this correct? Do I need to package for windows, linux and macos (I want consumers to download a pre-built packages not build themselves) in a single package? This is fine for now but might be a bit inefficient later on? Or am I missing something? The example given at https://github.com/jgsogo/conan-xbuild was not entirely clear for me. 

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by jgsogo** [comment]
Hi, @mjvk !

First of all, thanks for trying the new model and submitting this issue, feedback is very important in this early stage to tune the model and improve it. The new model _is not compatible_ with the old `os_build` and `arch_build` settings (it is trying to deprecate `xxx_build` settings), so try not to mix packages using those settings.

---

In a cross compiling scenario you should have a cross-compiler as a package, this binary should be identified (as any other binary) with the settings corresponding to the machine where it is going to run (the `host` settings).

### Package for the cross-compiler
So, if your settings are:

```ini
os: Neutrino
arch: x86_64
compiler: qcc
```


it means that your cross compiler is going to run in OS neutrino/x64 and it has been compiled using qcc.

I would have expected your cross-compiler to run in Windows/Linux/Mac and cross-compile to `target` Neutrino, so the settings for that package should have been:

```ini
os: Windows
arch: x86_64
compiler: Visual Studio
compiler.version: 16
build_type: Release
```

which means that it is binary that has been compiled using Visual Studio 16 and it is going to run in Windows/64. Something you create using `conan create my-crosscompiler/1.0 -s os=Windows -s compiler="Visual Studio" -s build_type=Release ....`  in your Windows machine. 

Probably, your cross-compiler recipe will have one or several options that will say something about the `target`, options like `target_system: ["Neutrino", ...]`, `target_arch: ["armv7", "armv8",...]`... whatever options make sense for your cross-compiler and **result in different binaries of your cross-compiler** (because they package different libraries from the target architecture)

So your cross compiler is a regular Conan package, with tons of combinations of settings and options that can create different binaries.

You can create this package using your Windows machine (the build machine where your compiler is running) and using just the `host` profile (Windows, Visual Studio 16, Release):

```
conan create my-crosscompiler/1.0 --profile=windows_profile
```

----

### Using the cross compiler

Then you will have the recipe for your application or library, in theory it could be any package you already have or any package from Conan Center. (I say _in theory_ because most of the times the recipe has to be prepared to use a cross compiler and the naïve approach with `cmake = CMake(); cmake.configure(); cmake.build()` won't work, but we are working on that).

Here you want to generate a binary that will run in Neutrino, so the settings for this binary would be `os=Neutrino,...` and you want to identify it also by the compiler used to generate it `compiler=my-crosscompiler, compiler.version=1.0`. Let's say you write all these settings into a profile called `neutrino_profile`:

```
os=Neutrino
arch=armv7
compiler=my-crosscompiler
compiler.version=1.0
build_type=Release
```

Now you want to run a command that:
 * cross-compiles:
   * runs in Windows (your `build` machine): you will need `--profile:build=windows_profile`
   * generate binaries for Neutrino (your `host` machine): you will need `--profile_host=neutrino_profile`
 * uses the package `my-crosscompiler/1.0` as cross-compiler: you need to declare this package as a build require of packages in the `host` context, so you need to add it to the profile host (to your `neutrino_profile`) (or hardcode it the recipe of `your-package`, but it is usually not desired as this recipe should work for other compilers too).

Let's go step by step:

 * the profile for your `host` (`--profile:host=neutrino_profile`) is quite straightforward:
   ```
   [settings]
   os=Neutrino
   arch=armv7
   compiler=my-crosscompiler
   compiler.version=1.0
   build_type=Release

   [build_requires]
   my-crosscompiler/1.0
   ```
 * your profile for your `build` context (`--profile:build=windows_profile`) is dedicated to the `my-crosscompiler` package: it runs in Windows and it should be able to select the proper binary from your cross-compiler

   ```
   [settings]
   os=Windows
   arch=x86_54
   compiler=Visual Studio
   compiler.version=16
   build_type=Release

   [options]
   my-crosscompiler:target_system=Neutrino
   my-crosscompiler:target_arch=armv7
   ```

   Remember that you can write the options in the command line, or maybe use profile composition to not mix settings about the `build` machine and options related to your cross-compiler.

 * The command to run it all will be:

   ```
   conan create <path/to/your-package> your-package/1.0@ --profile:host=neutrino_profile --profile:build=windows_profile
   ```

   Conan will create a graph with two profiles:

![image](https://user-images.githubusercontent.com/1406456/81267677-24a39700-9047-11ea-80b3-d45a17df6e05.png)

   your cross-compiler will live in the `build` context (Windows) and Conan will retrieve a binary that runs in your Windows machine, this binary will use the proper options to cross-compile to the `host` machine (Neutrino/armv7).
   The object instantiated by Conan from the `your-package` recipe will have available during the `build()` function call the environment defined in the `my-crosscompiler` package, this package should probably populate the `PATH` and maybe add some `CC` or `CXX` environment variables. Depending on your cross-compiler the logic in that `build()` method will be more or less straightforward.

----

I hope this helps you a little bit with this topic, if not, please blame me because of my poor explanation 😓 and I'll try to improve it. Also, if you feel like sharing with us your example we will be pleased to help you and maybe add it to our examples repository.

Hope this helps! Thanks!

**Comment by jgsogo** [comment]
Ah! the new member variables `settings_build` and `settings_target` can help you to know the other contexts that play a role in your execution, you can use them to check the options you are using in the `my-crosscompiler` package and maybe choose between `Release` and `Debug` without using an additional option (that will require 2x more binaries).

**Comment by mjvankampen** [comment]
Wow thanks for the extensive example! I used this crosscompiler before by using the *_build settings. I did not need to use options as the package was already identified by os, arch, os_build and arch_build. That is what I missed.

As target_build is not a property of a package I cannot distinguish between packages and need to add options correct?

My surprise was that you need to define target_arch and target_system in your build_profile, while you already define it in your host profile. So if I want to change build_profile I also need to make the same change in my host profile. This seems a bit less nice than the previous solution where you defined it only once (in your single profile). Although if you combine all targets in a single package it works fine as before by checking the settings_target, and of course error reporting works like a charm.

I will try your solution a bit and report back (with an example if I manage it).

My line to run this was:

conan export-pkg . -s:h arch=x86_64 -s:b arch=x86_64 -s:h os=Neutrino -s:b os=Windows -s:h compiler=qcc -s:h compiler.libcxx=cxx -s:h compiler.version=5.4 -s:h os.version=7.0

this would become

conan export-pkg . -s arch=x86_64 -s os=Windows 

To make it more specific I would add options using -o and check using settings_target if the options match with the target settings.

Ps. really happy with this change! Keep up the awesome work you guys are doing!

**Comment by jgsogo** [comment]
> Wow thanks for the extensive example! I used this crosscompiler before by using the *_build settings. I did not need to use options as the package was already identified by os, arch, os_build and arch_build. That is what I missed.
> 

Now the binaries for your cross-compiler can be identified using a full-profile... and recipes in Conan Center don't need to provide the _installer_ version. Now there are recipes like `libgettext` (regular settings "os", "compiler", "arch", "build_type") and `gettext` ("os_build", "arch_build"). With this new approach those two packages can be merged into a single one (they are actually the same).

> As target_build is not a property of a package I cannot distinguish between packages and need to add options correct?

If the binary for your cross-compiler is different depending on the target then, yes, you need to provide something to Conan so it computes a different package ID. If the binary is always the same (because it can target only one architecture, or it already includes everything needed for all the possible targets) then you don't need that option, and you can use in runtime the value of `settings_target` to know the target architecture your cross-compiler should use.

> 
> My surprise was that you need to define target_arch and target_system in your build_profile, while you already define it in your host profile. So if I want to change build_profile I also need to make the same change in my host profile. This seems a bit less nice than the previous solution where you defined it only once (in your single profile). Although if you combine all targets in a single package it works fine as before by checking the settings_target, and of course error reporting works like a charm.

You need to define them in the `profile_host`, that's needed, for sure. Then you need to define those options in the `build` context if they affect your package, Conan needs to know the package ID of your cross-compiler. As I said before, you need those options if the binaries are different, but you don't need them if the binary is the same, your cross-compiler can take those values from `self.settings_target` and decide the flags to use.

> 
> I will try your solution a bit and report back (with an example if I manage it).
> 
> My line to run this was:
> 
> conan export-pkg . -s:h arch=x86_64 -s:b arch=x86_64 -s:h os=Neutrino -s:b os=Windows -s:h compiler=qcc -s:h compiler.libcxx=cxx -s:h compiler.version=5.4 -s:h os.version=7.0
> 
> this would become
> 
> conan export-pkg . -s arch=x86_64 -s os=Windows
> 
> To make it more specific I would add options using -o and check using settings_target if the options match with the target settings.

If you are only running `export-pkg` then you are not compiling anything. You don't need the `build` context at all. Am I missing something?

> 
> Ps. really happy with this change! Keep up the awesome work you guys are doing!


Please, keep us updated
